app/lib/main/run_app.dart

```diff
@@ -117,46 +137,60 @@ Future runFlutterApp() async {
         einkommendeLinks: dynamicLinkBloc.einkommendeLinks,
         istGruppeBereitsBeigetreten: (publicKey) async =>
             await istSchonGruppeMitSharecodeBeigetreten(
-                sharezoneGateway, publicKey),
+          sharezoneGateway,
+          publicKey,
+        ),
       );
 
       gruppenBeitrittsTransformer.gefilterteBeitrittsversuche.listen(
-          beitrittsversuche.add,
-          onError: beitrittsversuche.addError,
-          cancelOnError: false);
+        beitrittsversuche.add,
+        onError: beitrittsversuche.addError,
+        cancelOnError: false,
+      );
 
-      UserGateway(references, currentUser).userStream.listen((user) {
+      userGateway = UserGateway(references, currentUser);
+      userGateway.userStream.listen((user) {
         if (user?.typeOfUser != null) {
           analytics.setUserProperty(
               name: 'typeOfUser', value: enumToString(user.typeOfUser));
         }
       });
+    } else {
+      // When the user signs out, we need to dispose the listeners and stream
+      // subscriptions inside the gateways. Otherwise, we we would cause a
+      // memory leak and receiving a permission denied error from Firestore
+      // because we would try to access the user data after the user signed out.
+      // This would result an instant fail of the integration tests.
+
+      if (userGateway != null) {
+        await userGateway.dispose();
+      }
+
+      if (sharezoneGateway != null) {
+        await sharezoneGateway.dispose();
+      }
```

Comment:
@Jonas-Sander I have managed to modify `SharezoneGateway`, `UserGateway`, `ConnectionsGateway`, `DataDocumentPackage` and `DataCombinedPackage` a bit so that we are able to dispose them correctly (not for the whole app but for the `listenToAuthStateChanged` listener). Therefore, we can remove is `isIntegrationTest` flag 🙂

Reply:
But i need to check manually that this doesn't break anything in the app.

Reply:
Didn't find any issues for now.

Reply:
Nit: Might it be more readable to do it as a pre-check via `if (user.typeOfUser == null)` instead of running it inside the `else` clause?

Also couldn't one do `userGateway?.dispose()`?